File: DataSets/DataSet-Old/LeetCodeTasksDevelop/LeetCodeTasks/3011_find-if-array-can-be-sorted/solutions/cpp/solution.cpp
#include <vector>
#include <unordered_map>
#include <algorithm>

class Solution {
public:
    bool canSortArray(std::vector<int>& nums) {
        std::unordered_map<int, std::vector<int>> bitMap;
        
        for (int num : nums) {
            int bitCount = __builtin_popcount(num);
            bitMap[bitCount].push_back(num);
        }
        
        std::vector<int> sortedNums;
        
        for (auto& [_, group] : bitMap) {
            std::sort(group.begin(), group.end());
            sortedNums.insert(sortedNums.end(), group.begin(), group.end());
        }
        
        std::sort(nums.begin(), nums.end());
        
        return sortedNums == nums;
    }
};

File: DataSets/DataSet-Old/LeetCodeTasksDevelop/LeetCodeTasks/3011_find-if-array-can-be-sorted/solutions/python3/solution.py
class Solution:
    def canSortArray(self, nums: list[int]) -> bool:
        from collections import defaultdict
        
        # Map to store numbers based on their bit count
        bit_map = defaultdict(list)
        
        # Populate the map with numbers grouped by their bit count
        for num in nums:
            bit_count = bin(num).count('1')
            bit_map[bit_count].append(num)
        
        # Now sort each list in the map and then flatten them
        # according to sorted bit counts
        sorted_list = []
        for bit_count in sorted(bit_map):
            bit_map[bit_count].sort()
            sorted_list.extend(bit_map[bit_count])
        
        # Compare the newly sorted list with the original list
        return sorted_list == sorted(nums)
